### PR TITLE
Add fault tolerance Streaming Dataset 2/n

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR addresses the comments from @awaelchli on this PR: https://github.com/Lightning-AI/lightning/pull/19049

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--19052.org.readthedocs.build/en/19052/

cc @borda

### CODE REVIEW

File: src/lightning/data/streaming/shuffle.py [hunk 1]
```diff
@@ -144,5 +144,5 @@ def get_chunks_and_intervals_per_ranks(self, distributed_env: _DistributedEnv, c
 
         return chunks_per_ranks, intervals_per_ranks
 
-    def __call__(self, array: np.ndarray, current_epoch: int, chunk_index: int) -> List[int]:
-        return np.random.RandomState(seed=self.seed + current_epoch + chunk_index).permutation(array).tolist()
+    def __call__(self, array: np.ndarray, epoch_shift: int, chunk_index: int) -> List[int]:
+        return np.random.RandomState(seed=self.seed + epoch_shift + chunk_index).permutation(array).tolist()
```
Comment (awaelchli) [inline src/lightning/data/streaming/shuffle.py:148]: renaming the variable won't help. My comment still applies. The random state will be the same if the sum of  (epoch_shift + chunk_index) is the same for different values of epoch_shift and chunk_index

Comment (awaelchli) [reply src/lightning/data/streaming/shuffle.py:148]: How about you just pass the sequence to the random state directly?

```suggestion
        return np.random.RandomState([self.seed, current_epoch, chunk_index]).permutation(array).tolist()
```

File: src/lightning/data/streaming/dataset.py [hunk 2]
```diff
@@ -284,55 +294,29 @@ def checkpoint(self, chunk_index: int) -> None:
...
-
             # 4. Move the file to its target position
-            shutil.move(tmp_checkpoint_path, checkpoint_path)
+            shutil.move(tmp_checkpoint_path, os.path.join(self.cache.checkpoint_rank_dir, "checkpoint.json"))
 
         self.last_time = time()
 
     def state_dict(self) -> Dict[str, Any]:
+        if is_in_dataloader_worker():
+            raise RuntimeError("The method `state_dict` is meant to be called only outside a DataLoader workers.")
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:304]:
```suggestion
            raise RuntimeError("The method `state_dict` should only be called in the main process.")
```

File: src/lightning/data/streaming/cache.py [hunk 3]
```diff
@@ -109,16 +114,17 @@ def filled(self) -> bool:
...
     def checkpoint_rank_dir(self) -> str:
-        checkpoint_rank_dir = os.path.join(self.checkpoint_dir, str(self.rank))
-        if not os.path.exists(checkpoint_rank_dir):
-            os.makedirs(checkpoint_rank_dir, exist_ok=True)
-        return checkpoint_rank_dir
+        checkpoint_rank_dir = os.path.join(self._cache_dir, "checkpoints", str(self.rank))
+        return self._try_create(checkpoint_rank_dir)
+
+    def _try_create(self, path: str) -> str:
+        if not os.path.exists(path):
```
Comment (awaelchli) [inline src/lightning/data/streaming/cache.py:125]: why is the check needed if we do `exist_ok=True` anyway? Is your intention to check for a file?

File: src/lightning/data/streaming/shuffle.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:148]: But now, it should be unique for every chunk every epoch.

Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:148]: As shift_epoch is `num_chunks * current_epoch`.

Comment (awaelchli) [reply src/lightning/data/streaming/shuffle.py:148]: Outside the function yes, but inside the function shift_epoch and index can be values that chosen such that they give the same value. It shouldn't be the responsibility of the caller to know the implementation detail. So my suggestion is to either move your change into the function or use the sequence as input to the random state.

File: src/lightning/data/streaming/cache.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/cache.py:125]: removed.

File: src/lightning/data/streaming/dataset.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:304]: done !

File: src/lightning/data/streaming/dataset.py [hunk 4]
```diff
@@ -413,6 +397,42 @@ def _string_to_datetime(item: str) -> datetime:
...
+        state = {}
+        obj = [state_dict]
+        torch.distributed.broadcast_object_list(obj, index, group=_group.WORLD)
+        state = obj[0]
+        state_dict_out.update(**state)
+        node_ranks.append(node_rank)
+    return state_dict_out
+
+
+def is_in_dataloader_worker() -> bool:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:432]: done !

File: src/lightning/data/streaming/shuffle.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:148]: Yes, you want this to be deterministic for a given epoch_shift and chunk_index.

Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:148]: This also simply random seed handling where we don't need to keep track of it nor to store it.

Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:148]: What do you think @awaelchli ?

**Review by awaelchli — APPROVED** [review]

**Review by rusenask — APPROVED** [review]
